File: apps/typescript/app-14-telemedicine/src/controllers/AuthController.ts
import { Request, Response } from "express";
import { AuthService } from "../services/AuthService";

export class AuthController {
  constructor(private readonly authService: AuthService) {}

  register = (req: Request, res: Response) => {
    const { username, password } = req.body;
    if (!username || !password) {
      return res.status(400).json({ message: "Username and password required." });
    }
    const user = this.authService.register(String(username), String(password));
    return res.json({ success: true, userId: user.id });
  };

  login = (req: Request, res: Response) => {
    const { username, password } = req.body;
    const session = this.authService.login(String(username ?? ""), String(password ?? ""));
    if (!session) {
      return res.status(401).json({ message: "Invalid credentials." });
    }

    res.cookie("token", session.token, {
      httpOnly: false,
      secure: false,
      maxAge: 7200000
    });
    return res.json({ success: true, user: session.user });
  };

  logout = (_req: Request, res: Response) => {
    res.clearCookie("token");
    return res.json({ success: true });
  };

  me = (req: Request, res: Response) => {
    const user = this.authService.requireUser(req.cookies?.token);
    if (!user) {
      return res.status(401).json({ message: "Access denied." });
    }
    return res.json({ user });
  };
}

File: apps/python/app-05-learning-mgmt/src/controllers/auth_controller.py
from flask import jsonify, request, session, make_response, render_template

from src.services.auth_service import AuthService


auth_service = AuthService()


def login():
    data = request.get_json() or {}
    user = auth_service.authenticate(data.get("username", ""), data.get("password", ""))
    if not user:
        return jsonify({"success": False, "message": "Invalid credentials"}), 401
    session["user_id"] = user["id"]
    session["username"] = user["username"]
    session["role"] = user["role"]
    return jsonify({"success": True, "user": {"username": user["username"], "role": user["role"]}})


def logout():
    session.clear()
    return jsonify({"success": True})


def get_me():
    user = auth_service.current_user(session)
    if not user:
        return jsonify({"message": "Unauthenticated"}), 401
    return jsonify({"username": user["username"], "role": user["role"]})


def dashboard_login():
    # VULNERABILITY A07: Dashboard session cookie set without httpOnly or secure flags
    data = request.get_json() or {}
    user = auth_service.authenticate(data.get("username", ""), data.get("password", ""))
    if not user:
        return jsonify({"success": False, "message": "Invalid credentials"}), 401
    session["user_id"] = user["id"]
    session["username"] = user["username"]
    session["role"] = user["role"]
    resp = make_response(jsonify({"success": True, "redirect": "/dashboard/student"}))
    resp.set_cookie("session", session.sid if hasattr(session, "sid") else "", httponly=False, secure=False)
    return resp
